File: language/python/concurrency/bounded_gather.py
"""
手写 bounded_gather。

考点：
- asyncio 并发限制：用 Semaphore 控制同时运行的协程数量。
- 收集所有结果，保持输入顺序。
- 支持超时和异常处理。
"""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable, Iterable
from typing import TypeVar

T = TypeVar("T")
R = TypeVar("R")


async def bounded_gather(
    coro_factory: Callable[[T], Awaitable[R]],
    items: Iterable[T],
    *,
    limit: int = 5,
    timeout: float | None = None,
    return_exceptions: bool = False,
) -> list[R]:
    """Run async tasks with bounded concurrency.

    Args:
        coro_factory: Async function applied to each item.
        items: Input sequence.
        limit: Maximum number of concurrent tasks.
        timeout: Optional timeout per task.
        return_exceptions: If True, exceptions are returned instead of raised.

    Returns:
        Results in the same order as ``items``.
    """
    if limit <= 0:
        raise ValueError("limit must be positive")

    semaphore = asyncio.Semaphore(limit)

    async def wrapped(item: T) -> R:
        async with semaphore:
            coro = coro_factory(item)
            if timeout is not None:
                coro = asyncio.wait_for(coro, timeout=timeout)
            return await coro

    tasks = [asyncio.create_task(wrapped(item)) for item in items]
    results = await asyncio.gather(*tasks, return_exceptions=return_exceptions)
    return results  # type: ignore[return-value]


if __name__ == "__main__":
    async def fetch(url: str) -> str:
        await asyncio.sleep(0.01)
        return f"data:{url}"

    async def main() -> None:
        urls = [f"url{i}" for i in range(10)]
        results = await bounded_gather(fetch, urls, limit=3)
        print(results)

    asyncio.run(main())

File: language/typescript/concurrency/boundedGather.ts
/**
 * 手写 boundedGather
 *
 * 考点：
 * - 用信号量限制并发数。
 * - 保持结果顺序。
 * - 支持 AbortSignal。
 */

export async function boundedGather<T, R>(
  items: Iterable<T>,
  fn: (item: T, signal?: AbortSignal) => Promise<R>,
  options: { limit?: number; signal?: AbortSignal } = {}
): Promise<R[]> {
  const { limit = 5, signal } = options;
  if (limit <= 0) throw new Error("limit must be positive");

  const itemsArray = Array.from(items);
  const results: R[] = new Array(itemsArray.length);
  let running = 0;
  let index = 0;

  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(signal.reason);
      return;
    }
    signal?.addEventListener("abort", () => reject(signal.reason), { once: true });

    function next(): void {
      while (running < limit && index < itemsArray.length) {
        const currentIndex = index++;
        running++;
        fn(itemsArray[currentIndex], signal)
          .then((result) => {
            results[currentIndex] = result;
          })
          .catch(reject)
          .finally(() => {
            running--;
            next();
          });
      }
      if (running === 0 && index >= itemsArray.length) {
        resolve(results);
      }
    }

    next();
  });
}

if (import.meta.url === `file://${process.argv[1]}`) {
  async function demo() {
    const results = await boundedGather(
      [1, 2, 3, 4, 5],
      async (n) => n * 2,
      { limit: 2 }
    );
    console.log(results); // [2, 4, 6, 8, 10]
  }
  demo();
}
